runtime/src/rent_collector.rs

```diff
@@ -147,14 +151,20 @@ impl RentCollector {
             return RentResult::LeaveAloneNoRent;
         }
 
-        let epoch_increment = match rent_due {
+        let new_rent_epoch = match rent_due {
             // Rent isn't collected for the next epoch
             // Make sure to check exempt status again later in current epoch
-            RentDue::Exempt => 0,
+            RentDue::Exempt => {
+                if self.update_rent_epoch_on_exempt {
+                    self.epoch
+                } else {
+                    account.rent_epoch() // leave unchanged
```

Comment:
:eyes:

Reply:
This is the key item in this pr. It turns out there are edge cases where rent_epoch is not updated and is updated depending on certain code paths. For example, a writable account in a successful tx WILL have its rent_epoch field updated. But, a failed tx where we collect a fee from the fee payer on failure does NOT have its rent_epoch field updated.